File: src/1306_JumpGameIII.java
class JumpGameIII {
  boolean[] seen;
  public boolean canReach(int[] arr, int start) {
    seen = new boolean[arr.length];
    return dfs(start, arr);
  }

  boolean dfs(int idx, int[] arr) {
    if (idx<0 || idx>=arr.length || seen[idx])
      return false;
    seen[idx] = true;
    return arr[idx]==0 ? true : (dfs(idx+arr[idx], arr) || dfs(idx-arr[idx], arr));
  }
}

File: src/1306_JumpGameIII.cpp
#include "utils.hpp"

class Solution {
  private:
    int n;

    bool dfs(int idx, vector<int>& arr, vector<bool>& seen) {
      if (idx<0 || idx>=n || seen[idx]) 
        return false;

      seen[idx] = true;
      if (arr[idx]==0) 
        return true;
      return dfs(idx+arr[idx], arr, seen) || dfs(idx-arr[idx], arr, seen);
    }
  public:
    bool canReach(vector<int>& arr, int start) {
      n = arr.size();
      vector<bool> seen(n);

      return dfs(start, arr, seen);
    }
};


int main() {
  UFUNC(Solution::canReach);
  return 0;
}